architecture/2024-08-applicationdeployment-controller.md

```diff
@@ -0,0 +1,370 @@
+
+The `radius-controller` will be responsible for reconciling ApplicationDeployment resources. When an ApplicationDeployment resource is created or updated, the controller will read the Bicep template and parameters from the ApplicationDeployment resource and send them to the `radius-bicep` service. The `radius-bicep` service will compile the Bicep and bicepparam files into ARM JSON and return the ARM JSON to the `radius-controller`. The `radius-controller` will then send the ARM JSON payload to UCP to deploy the resources.
+
+### Architecture Diagram
+<!--
+Provide a diagram of the system architecture, illustrating how different
+components interact with each other in the context of this proposal.
+
+Include separate high level architecture diagram and component specific diagrams, wherever appropriate.
+-->
+
+![Architecture Diagram](2024-08-applicationdeployment-controller/architecture.png)
+
+### Detailed Design
+
+<!--
+This section should be detailed and thorough enough that another developer
+could implement your design and provide enough detail to get a high confidence
+estimate of the cost to implement the feature but isn’t as detailed as the 
+code. Be sure to also consider testability in your design.
+
+For each change, give each "change" in the proposal its own section and
+describe it in enough detail that someone else could implement it. Cover
+ALL of the important decisions like names. Your goal is to get an agreement
+to proceed with coding and PRs.
+
+If there are alternatives you are considering please include that in the open
+questions section. If the product has a layered architecture, it's good to
+align these sections with the product's layers. This will help readers use
+their current understanding to understand your ideas.
+
+Discuss the rationale behind architectural choices and alternative options 
+considered during the design process.
+-->
+
+#### ApplicationDeployment Custom Resource Definition
+
+The `ApplicationDeployment` resource will be a new Kubernetes CRD that will be used to contain the Bicep template and parameters for deploying Radius resources. The CRD will have the following fields:
+
+- `spec.template`: The Bicep template file that defines the resources to deploy
+- `spec.parameters`: The Bicep parameters file that defines the parameters for the template
+- `status`: The status of the deployment, which will be one of the following values:
+  - `Pending`: The deployment is pending
+  - `Deployed`: The deployment has been completed
+  - `Missing Dependencies`: The deployment is missing dependencies
+
+#### `rad applicationdeployment generate` Command
+
+The `rad applicationdeployment generate` command will be a new command that will generate an ApplicationDeployment resource from a Bicep template and parameters file. The command will take the following arguments:
+
+- `-t, --template`: The path to the Bicep template file
+- `-p, --parameters`: The path to the Bicep parameters file
```

Comment:
Take a look at the options we support for `rad deploy` and how parameters are specified there.